src/lib/network_pool/snark_pool.ml

```diff
@@ -104,6 +104,65 @@ end
 
 module Make (Transition_frontier : Transition_frontier_intf) :
   S with type transition_frontier := Transition_frontier.t = struct
+  module Batcher : sig
+    type t [@@deriving sexp]
+
+    val create : Verifier.t -> t
+
+    val verify :
+         t
+      -> (Ledger_proof.t * Coda_base.Sok_message.t) list
+      -> bool Or_error.t Deferred.t
+  end = struct
+    type state =
+      | Waiting
+      | Verifying of
+          { out_for_verification:
+              (Ledger_proof.t * Coda_base.Sok_message.t) list
+          ; next_finished: bool Or_error.t Ivar.t sexp_opaque }
+    [@@deriving sexp]
+
+    type t =
+      { mutable state: state
+      ; queue: (Ledger_proof.t * Coda_base.Sok_message.t) Queue.t
+      ; verifier: Verifier.t sexp_opaque }
+    [@@deriving sexp]
+
+    let create verifier = {state= Waiting; queue= Queue.create (); verifier}
+
+    let call_verifier t ps = Verifier.verify_transaction_snarks t.verifier ps
+
+    (* When new proofs come in put them in the queue.
+       If state = Waiting, verify those proofs immediately.
+       Whenever the verifier returns, if the queue is nonempty, flush it into the verifier.
+    *)
+
+    let rec start_verifier t finished =
+      if Queue.is_empty t.queue then (
+        t.state <- Waiting ;
+        Ivar.fill finished (Ok true) )
```

Comment:
it took me a bit to realize that this was a reachable state. this comment would have prevented my confusion.

```suggestion
        (* we looped in the else after verifier finished but no pending work. *)
        t.state <- Waiting ;
        Ivar.fill finished (Ok true) )
```